broadcaster/_backends/redis.py

```diff
@@ -1,34 +1,50 @@
-import asyncio_redis
+import aioredis
+import asyncio
 import typing
-from urllib.parse import urlparse
 from .base import BroadcastBackend
 from .._base import Event
 
 
 class RedisBackend(BroadcastBackend):
     def __init__(self, url: str):
-        parsed_url = urlparse(url)
-        self._host = parsed_url.hostname or "localhost"
-        self._port = parsed_url.port or 6379
+        self.conn_url = url
+
+        self._pub_conn: typing.Optional[aioredis.Redis] = None
+        self._sub_conn: typing.Optional[aioredis.Redis] = None
+
+        self._msg_queue: typing.Optional[asyncio.Queue] = None
+        self._tasks: typing.List[asyncio.Task] = []
 
     async def connect(self) -> None:
-        self._pub_conn = await asyncio_redis.Connection.create(self._host, self._port)
-        self._sub_conn = await asyncio_redis.Connection.create(self._host, self._port)
-        self._subscriber = await self._sub_conn.start_subscribe()
+        self._pub_conn = await aioredis.create_redis(self.conn_url)
+        self._sub_conn = await aioredis.create_redis(self.conn_url)
+        self._msg_queue = asyncio.Queue()  # must be created here, to get proper event loop
 
     async def disconnect(self) -> None:
         self._pub_conn.close()
         self._sub_conn.close()
 
+        self._pub_conn = None
+        self._sub_conn = None
+        self._msg_queue = None
+
     async def subscribe(self, channel: str) -> None:
-        await self._subscriber.subscribe([channel])
+        channels = await self._sub_conn.subscribe(channel)
+        self._tasks.append(asyncio.create_task(self.reader(channels[0]), name=f"{channel} reader"))
 
     async def unsubscribe(self, channel: str) -> None:
-        await self._subscriber.unsubscribe([channel])
+        await self._sub_conn.unsubscribe(channel)
+
+        for task in self._tasks:
+            await task
```

Comment:
Wouldn't you have to do something like `task.cancel()` here?

Reply:
I was following the example from the aioredis docs: https://aioredis.readthedocs.io/en/v1.3.0/examples.html#pub-sub-example

What _should_ happen is that after the call to usubscribe, the channel will be inactive so in the reader function, the while loop will stop.

Reply:
However, in the `disconnect` function we may want to cancel all outstanding tasks...